037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+
+* [RFC #1][rfc-1], now defunct, is similar to this proposal but had a concept
+  of "spaces" baked into the interface. This concept has been decoupled from
+  the resource interface and will be re-introduced in a later RFC that is
+  compatible with v1 and v2 resources.
+  * **Recommended reading**: [this comment][rfc-1-comment] outlines the thought
+    process that led to this RFC.
+
+* [concourse/concourse#534](https://github.com/concourse/concourse/issues/534)
+  was the first 'new resource interface' proposal which pre-dated the RFC
+  process.
+
+
+## Motivation
+
+* Support for 'reusable tasks' in order to stop the shoehorning of
+  functionality into the resource interface:
+  [concourse/rfcs#7](https://github.com/concourse/rfcs/issues/7).
+
+* Support for creating multiple versions from `put`:
+  [concourse/concourse#2660](https://github.com/concourse/concourse/issues/2660)
+
+* Support for deleting versions:
+  [concourse/concourse#362](https://github.com/concourse/concourse/issues/362),
+  [concourse/concourse#524](https://github.com/concourse/concourse/issues/524)
+
+* Having resource metadata immediately available via check:
+  [concourse/git-resource#193](https://github.com/concourse/git-resource/issues/193),
+  [concourse/concourse#1714](https://github.com/concourse/concourse/issues/1714)
+
+* Unifying `source` and `params` as just `config` so that resources don't have
+  to care where configuration is being set in pipelines:
+  [concourse/git-resource#172](https://github.com/concourse/git-resource/pull/172),
+  [concourse/bosh-deployment-resource#13](https://github.com/concourse/bosh-deployment-resource/issues/13),
+  [concourse/bosh-deployment-resource#6](https://github.com/concourse/bosh-deployment-resource/issues/6),
+  [concourse/cf-resource#20](https://github.com/concourse/cf-resource/pull/20),
+  [concourse/cf-resource#25](https://github.com/concourse/cf-resource/pull/25),
+  [concourse/git-resource#210](https://github.com/concourse/git-resource/pull/210)
+
+* Make resource actions reentrant so that we no longer receive `unexpected EOF`
+  errors when reattaching to an in-flight build whose resource action completed
+  while we weren't attached:
+  [concourse/concourse#1580](https://github.com/concourse/concourse/issues/1580)
+
+* Support for showing icons for resources in the web UI:
+  [concourse/concourse#788](https://github.com/concourse/concourse/issues/788),
+  [concourse/concourse#3220](https://github.com/concourse/concourse/pull/3220),
+  [concourse/concourse#3581](https://github.com/concourse/concourse/pull/3581)
+
+* Standardize TLS configuration so every resource doesn't implement their own
+  way: [concourse/rfcs#9](https://github.com/concourse/rfcs/issues/9)
+
+
+## Glossary
+
+* **Prototype**: an implementation of the interface defined by this proposal,
+  typically packaged as an OCI container image.
+
+  Examples:
+
+  * a `git` prototype for interacting with commits and branches in a repo.
```

Comment:
A few comments which together raise the question of how these prototypes should be used. Defining objects might feel nice, but what does it give rise to? A pipeline, which guarantees that a green box at the right means "a pipeline-passing source code and related artifacts were most recently deployed" is extremely valuable to me. While I am already able to violate this by not using well-designed resource types, the prototypes abstraction seems to encourage me to do so. To avoid this, I just need to look clearly at what a pipeline is: its value is *implied* by the structural properties of the resource prototype. I would expect one of a few changes that I would immediately do is re-implement most of the resources I regularly use as children of a general "resource prototype" anyway so that the pipeline is still meaningful to me.

Reply:
Overall, my review might be addressed in large part by providing an example of another highly useful pairing, comparable to (Resource prototype, pipeline), that I might implement using this new capability.

Reply:
I think there may be some confusion underlying a few of these comments; this is an underlying protocol, and not a replacement for resources. None of this should change. Resource types would simply be reimplemented with the same protocol, and pipelines - and your interpreting of their meaning - would stay the same. The only difference is that things which *should not* be represented as resources no longer *have* to be.

Reply:
First of all, thanks as always for your extremely well-considered and patient thoughts, feedback, and questions. It is such a pleasure and charm to work with you and this project :) 

and in case it does not come across- i am very excited for this update.

Based on further clarifying conversations, i think i had an Aha! moment, around how many of these prototypes we might expect to be in circulation. Per some of your comments above, you might imagine the "resource" concept being separated (rather than conflated, as it is now) with the "notification" concept , by way of each being its own prototype. Off-hand, i would be surprised if there were more than, say, 10 generally useful prototypes at this level of abstraction. I also have difficulty seeing how "git" or "oci image" is sibling to, rather than child of, "resource", when thinking in this way.

The best explanation that comes to mind is that `git prototype` is a toy example and so its function is less clear to me than the real example of `notification/notifier prototype`. And so my confusion would just have arisen around these different examples, and my suggestion i suppose would be to focus on examples like `notification`. If not, could you maybe refer me to a usage of the `git` prototype that would be useful (currently, the example in your protocol doc is more of the how than the why)?

Reply:
No worries. Thanks for engaging on this - this is a pretty intricate problem space so it's super helpful for finding the parts that aren't totally clear. 🙂

I think what you're suggesting is that these sets of *interfaces* would likely not be many, and I agree. I think the number of interfaces will tend to correspond to the number of common automation workflows and patterns that exist in the software industry. This is probably a smallish number like you said, though I'm still finding [new instances of them](https://github.com/concourse/concourse/discussions/4949#discussioncomment-5681) now and then.

The number of concrete implementations of these interfaces will correspond to the number of distinct tools and services available, which is a much larger number, but still bounded. (There will likely be competing implementations as well.)

However keep in mind this is also to support packaging and running arbitrary code which is not really expected to conform to a particular interface. This makes the set of implementations effectively unbounded.

> I also have difficulty seeing how "git" or "oci image" is sibling to, rather than child of, "resource", when thinking in this way.

Hm, I wouldn't say the `git` prototype is a sibling of a resource, I'd say the `git` prototype *contains* a resource implementation - perhaps even multiple, as we've found that folks want to version based on commit but also based on tags. The `git` prototype can implement objects which support the verbs which make up a resource (`check`, `get`, `put`), but it might also support other verbs which can be done to other types of objects.

Maybe a better example would be a `github-pr` prototype, as it would likely implement three interfaces: a var source, a resource, and a notifier. The end goal is to support enumerating open PRs in a repo (via a var source), discovering and fetching the commits within a PR (via a resource), and updating the status of the PR on GitHub when the build completes (via a notifier).

I'll try to demonstrate that here and maybe work it into the RFC if it goes over well. I'm gonna make up a notation: `some-type{some:object}.some-message` means send the `some-message` message to the `{some:object}` object using the `some-type` prototype.

Note that objects are literally just JSON objects - it's up to the prototype to figure out what the object is describing and what messages the object supports based on the structure of the object itself.

Given a repo object, which only has the `repo` field, it could know how to `list` the set of open PRs against a repo:

```
github-pr{repo:concourse/concourse}.list
=> {pr:123} {pr:124} ...
```

With #29, this response would be used to configure a pipeline for each pull request, using `pr: 123` as a var.

Each pipeline would configure a resource like so, configuring itself as a notifier (this is where you'd otherwise configure a Slack notifier or whatever):

```yaml
resources:
- name: pr
  type: github-pr
  notifiers: [self]
  source:
    repo: concourse/concourse
    pr: 123
```

(Sidebar: an alternative here could be to use the `git` resource prototype to track the PR branch, so as to not duplicate code between `github-pr` and `git`, while defining a separate `github-pr` notifier for the `git` resource. But for the sake of this demonstration let's say it's all baked into one prototype.)

Concourse would send `check` to this object to find commits:

```
github-pr{repo:concourse/concourse,pr:123}.check
=> {ref:commit-1} {ref:commit-2}
```

...And when a build runs, Concourse would `get` a specific commit object with the `get` step, which would run...:

```
github-pr{repo:concourse/concourse,pr:123,ref:commit-1}.get
=> {ref:commit-1} (and bits are fetched)
```

...and when the build completes, the commit object would be extended with build status information, and the notifier would be run, updating the PR status on GitHub:

```
github-pr{repo:concourse/concourse,pr:123,ref:commit-1,build:{id:123,status:"succeeded"}}.notify
=> (no response needed)
```

So in total, this single `github-pr` prototype implements a set of objects which together support three interfaces:

1. The repo object supports the `list` message, and so it could be used with #29 to fan-out across all PRs and set one pipeline for each.
1. The PR object supports `check` and returns commit objects which support `get`, allowing it to be used as a resource in the pipeline.
1. The commit status object supports the `notify` message, tying it in to #28 (which still needs to be fleshed out).